gprofiler/profilers/profiler_base.py

```diff
@@ -119,41 +156,131 @@ def _profiling_error_stack(
-                    )
-                    result = ProfileData(
-                        self._profiling_error_stack("error", "process went down during profiling", comm), None, None
-                    )
-                except Exception as e:
-                    logger.exception(f"{self.__class__.__name__}: failed to profile process {pid} ({comm})")
-                    result = ProfileData(
-                        self._profiling_error_stack("error", f"exception {type(e).__name__}", comm), None, None
-                    )
-
-                results[pid] = result
+            return self._wait_for_profiles(futures)
+
+
+class SpawningProcessProfilerBase(ProcessProfilerBase):
+    """
+    Enhances ProcessProfilerBase with tracking of newly spawned processes.
+    """
+
+    _SCHED_THREAD_INTERVAL = 0.1
+    _BACKOFF_INIT = 0.1
+    # so we wait up to 1.5 seconds
+    _BACKOFF_MAX = 0.8
+
+    def __init__(self, frequency: int, duration: int, stop_event: Optional[Event], storage_dir: str):
+        super().__init__(frequency, duration, stop_event, storage_dir)
+        self._submit_lock = Lock()
+        self._threads: Optional[ThreadPoolExecutor] = None
+        self._start_ts: Optional[float] = None
+        self._enabled_proc_events = False
+        self._futures: Dict[Future, Tuple[int, str]] = {}
+        self._sched = sched.scheduler()
+        self._sched_stop = False
+        self._sched_thread = Thread(target=self._sched_thread_run)
+
+    def _should_profile_process(self, pid: int) -> bool:
+        raise NotImplementedError
+
+    def _notify_selected_processes(self, processes: List[Process]) -> None:
+        # TODO ensure PIDs in _proc_exec_callback don't intersect with "processes"?
+        # now we start watching for new processes.
+        self._start_profiling_spawning()
 
+    @property
+    def _is_profiling_spawning(self) -> bool:
+        return self._threads is not None
+
+    def _start_profiling_spawning(self) -> None:
+        with self._submit_lock:
+            self._start_ts = time.monotonic()
+            self._threads = ThreadPoolExecutor()
+
+    def _stop_profiling_spawning(self) -> None:
+        with self._submit_lock:
+            self._start_ts = None
+            assert self._threads is not None
+            self._threads.shutdown()  # waits (although - all are done by now)
+            self._threads = None
+
+    def _proc_exec_callback(self, tid: int, pid: int) -> None:
+        with contextlib.suppress(NoSuchProcess):
+            self._sched.enter(self._BACKOFF_INIT, 0, self._check_process, (Process(pid), self._BACKOFF_INIT))
+
+    def start(self) -> None:
+        super().start()
+
+        self._sched_thread.start()
+
+        try:
+            register_exec_callback(self._proc_exec_callback)
+        except Exception:
+            logger.warning("Failed to enable proc_events listener for executed processes", exc_info=True)
+        else:
+            self._enabled_proc_events = True
```

Comment:
This also exists in `JavaProfiler`, so it's overwriting this one. If the Base manages to register its callback but `JavaProfiler` doesn't (or vice versa), then it'll be in inconsistent state.

Reply:
> Personally I think this is one of the reasons to use more compositions over inheritance here.

Makes sense. Next PR